Consistently get extensions from FileTypes.h

jrose-apple opened the pull request:
...instead of sometimes hardcoding them and sometimes using Strings.h. ~The exceptions are the libraries that sit below Frontend; these can continue using strings.~

There's one semantic change here, which is that LLVM IR files mistakenly had the extension "ir" instead of "ll" before. Nothing was actually depending on this until now, though.

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test

gottesmm commented:
Seems reasonable to me. I haven't looked through all of the details though.

davidungar reviewed (commented):
This is great stuff! Of all my comments, it's the ones about symmetry that strike me hardest. If it has to be this way, how about a comment somewhere explaining why all the type extensions can't be in one place?

jrose-apple commented on include/swift/Frontend/Types.def:
```diff
@@ -46,13 +46,15 @@ TYPE("object",              Object,                    "o",               "")
 TYPE("raw-sil",             RawSIL,                    "sil",             "")
 TYPE("raw-sib",             RawSIB,                    "sib",             "")
-TYPE("llvm-ir",             LLVM_IR,                   "ir",              "")
+TYPE("llvm-ir",             LLVM_IR,                   "ll",              "")
```
Yep, as commented above "ir" was a mistake but we weren't relying on it for anything until now.

davidungar replied:
Sorry-missed that comment.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -718,7 +718,7 @@ createStatsReporter(const llvm::opt::InputArgList *ArgList,
     InputName = Inputs[0].second->getSpelling();
   }
-  StringRef OutputType = file_types::getTypeTempSuffix(OI.CompilerOutputType);
+  StringRef OutputType = file_types::getExtension(OI.CompilerOutputType);
```
Fair enough.

jrose-apple commented on include/swift/Strings.h:
```diff
@@ -21,15 +21,6 @@ constexpr static const char SERIALIZED_MODULE_EXTENSION[] = "swiftmodule";
-constexpr static const char SIB_EXTENSION[] = "sib";
-/// The extension for LLVM IR files.
-constexpr static const char LLVM_BC_EXTENSION[] = "bc";
-constexpr static const char LLVM_IR_EXTENSION[] = "ll";
```
I have an okay reason. :-/ The remaining constants here are ones that are used in libraries below Frontend (in layering terms). I'd prefer not to put them *all* here because I think the `file_types` abstractions are better, but it does make an ugly asymmetry here.

Another option would be to go a step further and sink the information in `file_types` down to Basic, which means they could be included from anywhere.

davidungar replied:
I *like* that option: sinking the info down to Basic! It's a nice unification which I suspect will pay off in the future!

jrose-apple committed:
Rename file_types::getTypeTempSuffix to getExtension

Let's be honest: this isn't just for temporary files. Also, it's not
really a "suffix" if it doesn't include the leading dot.

No functionality change.

jrose-apple committed:
Consistently get extensions from swift/Frontend/FileTypes.h

...instead of sometimes hardcoding them and sometimes using Strings.h.
The exceptions are the libraries that sit below Frontend; these can
continue using strings.

jrose-apple committed:
Move swift/Frontend/FileTypes.h down to Basic

The next commit will take advantage of this, but this is just a
mechanical change.

jrose-apple committed:
Consistently get extensions from swift/Basic/FileTypes.h (part 2)

Remove the last few literal extension strings from Strings.h in favor
of the file_types APIs, and use those APIs in a few more places.

jrose-apple committed:
Minor improvements to file_types

No intended functionality change.

jrose-apple commented:
@swift-ci Please test

davidungar reviewed (approved):
Very nice! Love the symmetry. Would still want the consts, but it's not a big deal.
